### PR TITLE
refactor: replace `pkg/errors` by golang built-in errors

### LINKED ISSUE #2009: Switch from pkg/errors package to Go's errors package
Now that go supports erorrs.Is and other nice functionality that we previously relied on pkg/errors to do for us, let's get off of that and use the built-in errors package.

As I work on new features, I'll update as I go but it would be great if we could do a sweep get everything consistent.

### PR DESCRIPTION
**Pull request by epsxy**
👋🏻 Hey, I've seen that issue listed in your issues and I tried to start something. This is just a WIP as I wanted to check if this was still something you wanted to do and if you aligned with the spirit of the refactor.

# What does this change

This PR is a WIP to start to remove a couple of `pkg/errors` dependencies by the usage of the built-in golang `errors` and `fmt` packages.

Closes #2009.

# Notes for the reviewer

Former dependencies have been replaced following these rules:

- `errors.New(...)` -> `errors.New(...)` (from pkg/errors to built-in)
- `errors.Wrapf(err, "...", ...)` -> `fmt.Errorf("...: %w", ..., err)`
- `errors.Wrap(err, "...")` -> `fmt.Errorf("...: %w", err)`
- `errors.New(fmt.Sprintf(...))` -> `fmt.Errorf(...)`
- `errors.Errorf(...)` -> `fmt.Errorf(...)`
- `errors.Cause()` -> `errors.Unwrap()`

Don't hesitate to discuss these choices if you have other ideas, that's the main reason why I am submitting the PR early `.Wrap` doesn't exist in built-in functions, but I though wrapping them manually with `fmt.Errorf("...: %w", err)` was enough.

⚠️ The main thing to be quite careful about, is the former usages of `.Wrap` or `.Wrapf`, which weren't wrapped under a `err != nil` check. In that case we need to convert that statement into the new `fmt.Errorf` one **along with** nesting it into an error nil check.

# Reviewer Checklist

* Comment with /azp run test-porter-release if a magefile or build script was modified
* Comment with /azp run porter-integration if it's a non-trivial PR

### CODE REVIEW

**Comment by carolynvs** [comment]
@epsxy The substitutions you listed in the PR is exactly what I'm looking for and yes, we are still quite interested in getting rid of pkg/errors and using native wrapping just like you are with the `%w` formatting directive. 👍 

Since you aren't done yet, I'll hold off on reviewing further until you give me the go ahead. Thanks for taking this on!

File: pkg/tracing/traceLogger.go
Comment (epsxy) [inline pkg/tracing/traceLogger.go:204]: 👀 Careful review is needed here, as it's a non-trivial update. I have replaced the `pkg/errors` search for the closest error in the stack with recursive calls to `errors.Unwrap`.

File: pkg/porter/upgrade.go
Comment (epsxy) [inline pkg/porter/upgrade.go:76]: 👀 This is not the only place where an error was returned and not checked. I was wondering if we needed to add some more checks. I think most of the time I didn't added some to avoid mutating the logic and failing in places where we didn't failed in the past. So here, I've just ignored it the value returned.

**Comment by epsxy** [comment]
@carolynvs Everything should been replaced now, although I'll review it myself file per file tomorrow. Seems like the integration tests didn't run in the latest commits, I hope everything is alright (because I'm not able to run them locally). 

Some integration tests can still fail for example because I made small adjustments sometimes to matches Go lints (not starting error messages with uppercase, not finishing error messages with dots, etc).

File: pkg/porter/upgrade.go
Comment (carolynvs) [reply pkg/porter/upgrade.go:76]: We are adding in lint/vet checks that will catch these in the future and #2176 is fixing any problems in the current code base. So you don't need to worry about fixing that problem in this PR. 👍

**Comment by carolynvs** [comment]
Our CI pipeline was having trouble yesterday and wasn't triggering for a couple hours in the afternoon. I merged some more PRs this morning and there's a merge conflict that will prevent further builds from running on your PR.

I'll start reviewing what's here though since there's plenty enough to keep me busy while you resolve the merge conflicts.

File: go.mod
Comment (carolynvs) [inline go.mod:176]: 🎉

File: pkg/cnab/cnab-to-oci/registry.go
Comment (carolynvs) [inline pkg/cnab/cnab-to-oci/registry.go:199]: oops, lost a word here

```suggestion
		return false, fmt.Errorf("could not list images: %w", err)
```

File: pkg/encoding/encoding.go
Comment (carolynvs) [inline pkg/encoding/encoding.go:67]: Let's only return the first argument when there isn't an error
```suggestion
		return nil, fmt.Errorf("error marshaling to %s: %w", format, err)
```

File: pkg/manifest/manifest.go
Comment (carolynvs) [inline pkg/manifest/manifest.go:988]:
```suggestion
		return nil, fmt.Errorf("could not read from url %s: %w", path, err)
```

File: pkg/mixin/query/manifest_generator.go
Comment (carolynvs) [inline pkg/mixin/query/manifest_generator.go:36]:
```suggestion
		return nil, fmt.Errorf("could not marshal mixin build input for %s: %w", mixinName, err)
```

**Review by carolynvs — COMMENTED** [review]
I'm going to do this in batches so that I don't miss anything when I review. This is a go-over of the first "few" files. 😁

**Comment by epsxy** [comment]
> I'm going to do this in batches so that I don't miss anything when I review. This is a go-over of the first "few" files. 😁

🎉 👍🏻 Btw if you want to delay your review, I'll review it myself in a couple of hours, just to catch the basic stuff, like the missing words you've noticed. It may save you some time if I manage to catch trivial mistakes.

**Comment by carolynvs** [comment]
Sure I'll wait and review the rest after I get the 👍 from you

File: pkg/porter/parameters.go
Comment (epsxy) [inline pkg/porter/parameters.go:404]: ⚠️ FIXME: `return nil, fmt.Errorf()`

File: pkg/porter/schema.go
Comment (epsxy) [inline pkg/porter/schema.go:133]: 🛑 ⚠️ 👀 Imo (but maybe I'm wrong because I dont know the logic very well), there was an issue here. The code was 
```
actionItemSchema, ok := actionSchemas[action]["items"].(jsonSchema)
if err != nil && p.Debug {
    ...
}
```

I think we needed to use `ok` instead of `err`. Because previously, `ok` was never used nor checked. If it wasn't a mistake, we should replace `ok` by `_`.

File: pkg/porter/schema.go
Comment (epsxy) [inline pkg/porter/schema.go:140]: 🛑 ⚠️ FIXME Same here? (I didn't fixed that place though).

File: pkg/porter/schema.go
Comment (epsxy) [inline pkg/porter/schema.go:156]: 🛑 ⚠️ Same here (I fixed it).

File: pkg/porter/upgrade.go
Comment (epsxy) [reply pkg/porter/upgrade.go:76]: ⚠️ **FIXME** (Note  to myself)

File: pkg/portercontext/context.go
Comment (epsxy) [inline pkg/portercontext/context.go:457]: 👀 We can probably do:
```suggestion
				return err
```

Comment (epsxy) [reply pkg/portercontext/context.go:457]: ⚠️ **FIXME**, I think we have to return here. Note to myself: add a `return nil` after the error check, in `IsDir()`.

File: pkg/portercontext/helpers.go
Comment (epsxy) [inline pkg/portercontext/helpers.go:265]: ℹ️ Was suggested by the linter

File: pkg/runtime/runtime-manifest.go
Comment (epsxy) [inline pkg/runtime/runtime-manifest.go:198]: ⚠️ **FIXME**: remove comment (🤯 misclick keyboard shortcut probably)

File: pkg/runtime/runtime-manifest.go
Comment (epsxy) [inline pkg/runtime/runtime-manifest.go:471]: ⚠️ FIXME add nil check.

File: pkg/storage/parameterset_provider_helpers.go
Comment (epsxy) [inline pkg/storage/parameterset_provider_helpers.go:60]:
```suggestion
		return nil, fmt.Errorf("error reading %s as a parameter set: %w", path, err)
```
We probably want that too.

File: pkg/storage/plugin_adapter.go
Comment (epsxy) [inline pkg/storage/plugin_adapter.go:182]: ℹ️ `err` was not used.

**Review by epsxy — COMMENTED** [review]
First round with a couple comments 💬. @carolynvs I will fix the most blatant/obvious typos and mistakes. Some of my comments spotted some behavior changes which - I hope - are responsible for the integration/smoke test failing. I'm letting all this comments opened, so that you can have a look at my notes, having an extra pair of eyes can't hurt.

File: pkg/mixin/query/query.go
Comment (epsxy) [reply pkg/mixin/query/query.go:113]: 🆗 not needed, the nil check is above imo.

File: pkg/mixin/query/query.go
Comment (epsxy) [reply pkg/mixin/query/query.go:126]: 🆗 not needed, the nil check is above imo.

File: pkg/pkgmgmt/client/runner.go
Comment (epsxy) [reply pkg/pkgmgmt/client/runner.go:41]: ℹ️ Not needed, nil check above.

File: pkg/porter/parameters.go
Comment (epsxy) [reply pkg/porter/parameters.go:404]: not needed

File: pkg/runtime/runtime-manifest.go
Comment (epsxy) [reply pkg/runtime/runtime-manifest.go:471]: Not needed.

File: pkg/storage/parameterset_provider_helpers.go
Comment (epsxy) [reply pkg/storage/parameterset_provider_helpers.go:60]: not possible

File: pkg/build/dockerfile-generator.go
Comment (epsxy) [reply pkg/build/dockerfile-generator.go]: 7fbc9f45

File: pkg/pkgmgmt/client/filesystem.go
Comment (epsxy) [reply pkg/pkgmgmt/client/filesystem.go]: 9488201a

File: pkg/portercontext/context.go
Comment (epsxy) [reply pkg/portercontext/context.go:457]: e16dba14

**Comment by epsxy** [comment]
@carolynvs (sorry for the double ping) I think you can review. I'm not sure why the CI is failing yet (though it's **_less_** failing than before since I fixed the smoke tests tonight I think), but that may be related to one of my open comment/issue. I'll dig that tomorrow.

File: pkg/runtime/dependencies.go
Comment (epsxy) [inline pkg/runtime/dependencies.go]: Okay, this is the reason some tests fail. Fixed in this commit: afff47f9

File: pkg/tracing/traceLogger.go
Comment (epsxy) [reply pkg/tracing/traceLogger.go:204]: Okay, I think this is the origin of one failing integration test [here](https://dev.azure.com/getporter/porter/_build/results?buildId=4588&view=logs&j=c892508d-390a-5765-0ad0-b341c7fd0dd4&t=99b6391b-efca-5f28-e88f-b44afecfc894&l=10423):
```
=== RUN   TestPlugins_CatchStderr/plugin_throws_an_error
    pluggable_integration_test.go:59: 
        	Error Trace:	pluggable_integration_test.go:59
        	Error:      	Error message not equal:
        	            	expected: "could not connect to the secrets.testplugin.vault plugin: Unrecognized remote plugin message: \n\nThis usually means that the plugin is either invalid or simply\nneeds to be recompiled to support the latest protocol.: i am just test plugin. i don't function\n"
        	            	actual  : "Unrecognized remote plugin message: \n\nThis usually means that the plugin is either invalid or simply\nneeds to be recompiled to support the latest protocol."
        	Test:       	TestPlugins_CatchStderr/plugin_throws_an_error
```

We Unwrap the error differently and the error message is not the same anymore. The new one is less useful. We are failing [here](https://github.com/getporter/porter/blob/afff47f90cd0e8e08fd06db610304fc8e264813a/pkg/plugins/pluggable/connection.go#L157-L165), and we end up displaying the default error message returned by `go-plugin` which is [here](https://github.com/hashicorp/go-plugin/blob/5dee41c45dcec4c3b00c6d227bec8e5ebea56a8a/client.go#L697-L702).

So we have to fix what I did here and format the error message correctly, like before I guess. I'm not sure how I will do that yet, I'll try to debug that on a playground to see how to Unwrap properly errors "recursively".

**Comment by carolynvs** [comment]
Sorry about the merge conflicts, I'm going to do my best to get your PR merged ASAP since it touches so many files and it's going to get a conflict with pretty much every PR that we merge before yours.

File: pkg/porter/upgrade.go
Comment (carolynvs) [inline pkg/porter/upgrade.go:75]:
```suggestion
	if err = p.applyActionOptionsToInstallation(ctx, &i, opts.BundleActionOptions); err != nil {
	  return err
	}
```

File: pkg/portercontext/context.go
Comment (carolynvs) [inline pkg/portercontext/context.go:445]: Let's just return err here. I'm not quite sure what I was going for with the WithStack.
```suggestion
			return err
```

File: pkg/portercontext/context.go
Comment (carolynvs) [reply pkg/portercontext/context.go:457]: Agreed. Anywhere I was using errors.WithStack I believe we can just go with using the original error, and can skip the %w directive.

File: pkg/portercontext/context.go
Comment (carolynvs) [inline pkg/portercontext/context.go:470]:
```suggestion
		return err
```

File: pkg/portercontext/context.go
Comment (carolynvs) [inline pkg/portercontext/context.go:475]:
```suggestion
		return err
```

File: pkg/portercontext/context.go
Comment (carolynvs) [inline pkg/portercontext/context.go:480]:
```suggestion
		return err
```

File: pkg/portercontext/helpers.go
Comment (carolynvs) [reply pkg/portercontext/helpers.go:265]: I think once you update with the most recent changes on release/v1, this is already fixed by the other PR (#2176)

File: pkg/runtime/runtime-manifest.go
Comment (carolynvs) [reply pkg/runtime/runtime-manifest.go:198]: This should return the error.

File: pkg/storage/plugin_adapter.go
Comment (carolynvs) [reply pkg/storage/plugin_adapter.go:182]: Can you have it check and return the error when not nil?

File: pkg/tracing/traceLogger.go
Comment (carolynvs) [reply pkg/tracing/traceLogger.go:204]: This change in behavior is because of another PR that recently went in (#2185) that changed how the error you are handling is generated. You're right we don't need the extra logic anymore to unwrap the root error and can remove the commented out unwrap calls.

**Review by carolynvs — CHANGES_REQUESTED** [review]
This is super close! I'll get this merged as soon as its fixed up so that we don't hit more merge conflicts.

**Comment by carolynvs** [comment]
I'll squash when I merge, so don't worry about cleaning up the commit history. 👍

File: CONTRIBUTORS.md
Comment (carolynvs) [inline CONTRIBUTORS.md:74]: 🎉

**Comment by epsxy** [comment]
@carolynvs I fixed everything I think, along adding myself at the bottom of `CONTRIBUTORS.md`! I think we're good o/

**Review by carolynvs — APPROVED** [review]
OMG Thank you so much for this ginormous patch! It's great to finally be off of pkg/errors, this was incredibly helpful. 💯

**Pull request merged**
